Source/Dafny/Plugin.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Dafny.Plugins;
+
+namespace Microsoft.Dafny;
+
+/// <summary>
+/// This class is internal to Dafny. It wraps an assembly and an extracted configuration from this assembly,
+/// The configuration provides the methods to parse command-line arguments and obtain Rewriters 
+/// </summary>
+public class Plugin {
+  public Configuration Configuration;
+
+  private readonly string path;
+  private readonly Assembly assembly;
+  private System.Type[] rewriterTypes;
+
+  class AutomaticConfiguration : Configuration {
+    private Func<ErrorReporter, Rewriter>[] rewriters;
+    public AutomaticConfiguration(Func<ErrorReporter, Rewriter>[] rewriters) {
+      this.rewriters = rewriters;
+    }
+
+    public override Rewriter[] GetRewriters(ErrorReporter errorReporter) {
+      return rewriters.Select(funcErrorReporterRewriter =>
+        funcErrorReporterRewriter(errorReporter)).ToArray();
+    }
+  }
+
+  public static IEnumerable<System.Type> GetConfigurationsTypes(Assembly assembly) {
+    return assembly.GetTypes()
+      .Where(t => t.IsAssignableTo(typeof(Configuration)));
+  }
+
+  public Plugin(string pluginPath, string[] args, ErrorReporter reporter) {
```

Comment:
Given that the constructor does IO, consider making the constructor private and introducing a `static Plugin Load(...)` method instead to make it clear to the caller that IO work is done.